pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/executor/LeafSchedulerService.java

```diff
@@ -92,11 +99,43 @@ public void cancel(long requestId) {
     }
   }
 
+  @Override
+  public void awaitDataAvailable(OpChainId opChainId, long timeoutMs) {
+    OpChainStatus status = _opChainStatusMap.computeIfAbsent(opChainId, k -> new OpChainStatus());
+    status.awaitStatus(timeoutMs);
+  }
+
+  @Override
+  public void setDataAvailable(OpChainId opChainId) {
+    OpChainStatus status = _opChainStatusMap.computeIfAbsent(opChainId, k -> new OpChainStatus());
+    status.setDataAvailable();
+  }
+
   private void closeOpChain(OpChain opChain) {
     opChain.close();
   }
 
   private void cancelOpChain(OpChain opChain, Throwable t) {
     opChain.cancel(t);
   }
+
+  private static class OpChainStatus {
+    private final BlockingQueue<Boolean> _queue = new ArrayBlockingQueue<>(1);
```

Comment:
I would suggest to use a lock and condition here instead. I'm not sure, but it may provide better performance.